go/def.bzl

```diff
@@ -385,6 +430,127 @@ go_test = rule(
     test = True,
 )
 
+def cgo_library(name, srcs,
+                # TODO(yugui) Automatically resolve pkgname from srcs at runtime
+                pkgname,
+                toolchain="//go/toolchain:toolchain",
+                go_tool="//go/toolchain:go_tool",
+                copts=[],
+                linkopts=[]):
+  """Declares a cgo-enabled go library.
+
+  Args:
+    name: A unique name for this rule.
+    srcs: Source files to build with cgo.
+      
+      They can be .go files which contains `import "C"` or anything allowed
+      in srcs attribute of cc_library.
+    copts: Add these flags to the C++ compiler.
+    linkopts: Add these flags to the C++ linker.
+
+  NOTE:
+    `srcs` cannot contain pure-Go files which does not contain `import "C"`.
+    So you need to define another `go_library` when you need to build a go
+    package with both cgo-enabled and pure-Go sources.
+
+    ```
+    cgo_library(
+      name = "cgo-enabled",
+      srcs = ["cgo-enabled.go", "foo.cc", "bar.S", "baz.a"],
+    )
+
+    go_library(
+      name = "go_default_library",
+      srcs = ["pure-go.go"],
+      library = ":cgo_enabled",
+    )
+    ```
+  """
+  out_dir = "%s.cgo.dir" % name
+  pkg_depth = PACKAGE_NAME.count('/') + 1
+
+  go_srcs = [s for s in srcs if s.endswith('.go')]
+  c_srcs = [s for s in srcs if not s in go_srcs]
+  thunks = [cgo_thunk(s, out_dir) for s in go_srcs]
+  c_exports = [out_dir + "/" + s for s in ("_cgo_export.c", "_cgo_export.h")]
+  c_dummy = out_dir + "/_cgo_main.c"
+  gotypes = out_dir + "/_cgo_gotypes.go"
+
+  cgogen_outs = ([t.go_src for t in thunks] + [t.c_src for t in thunks] +
+                 c_exports + [c_dummy, gotypes])
+
+  native.genrule(
+      name = "%s.cgogen" % name,
+      srcs = srcs,
+      outs = cgogen_outs,
+      cmd = " && ".join([
+          "go_tool_dir=$$(cd $$(dirname $(location {go_tool})) && pwd)",
+          "go_tool=$$go_tool_dir/$$(basename $(location {go_tool}))", 
+          "objdir=$$(cd $(@D)/{out_dir} && pwd)",
+          "export GOROOT=$$go_tool_dir/..",
+          "export CGO_LDFLAGS='" + " ".join(linkopts) + "'",
+          "mkdir -p $$objdir",
+          "cd {pkgname}",
+          ("$$go_tool tool cgo -objdir $$objdir " + " ".join(go_srcs)), 
+          "rm -f $$objdir/_cgo_.o $$objdir/_cgo_flags",
+      ]).format(go_tool=go_tool, out_dir=out_dir, pkgname=PACKAGE_NAME),
```

Comment:
pkgname=pkgname, no? 

PACKAGE_NAME is the directory leading up to the BUILD file.